src/cmd/run-smoketest/main.go

```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		return nil, err
+	}
+
+	commands := []string{}
+	scanner := bufio.NewScanner(&buf)
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		ofs := 0
+		for _, f := range fields {
+			spl := strings.SplitN(f, "=", 2)
+			if len(spl) == 2 {
+				ofs++
+			} else {
+				break
+			}
+		}
+		fields = fields[ofs:]
+
+		if filepath.Base(fields[0]) == "tinygo" && (fields[1] == "build" || fields[1] == "version") {
+			commands = append(commands, line)
+		} else if strings.HasPrefix(line, "#") {
+			commands = append(commands, line)
+		}
+	}
+
+	return commands, nil
+}
+
+func run(commands []string, threads int) error {
+	ch := make(chan string, 1000)
+	o := NewOchan(ch, 100)
+
+	go func() {
+		limit := make(chan struct{}, threads)
+		var eg errgroup.Group
+
+		for _, command := range commands {
+			command := command
+			select {
+			case limit <- struct{}{}:
+				och := o.GetCh()
+				eg.Go(func() error {
+					var err error
+
+					fields := strings.Fields(command)
+					ofs := 0
+					for _, f := range fields {
+						spl := strings.SplitN(f, "=", 2)
+						if len(spl) == 2 {
+							ofs++
+						} else {
+							break
+						}
+					}
+					fields = fields[ofs:]
+
+					if filepath.Base(fields[0]) == "tinygo" && fields[1] == "build" {
+						err = buildAndmd5sum(command, och)
+					} else if strings.HasPrefix(command, "#") {
+						och <- fmt.Sprintf("%s\n", command)
+					} else {
+						err = runCommand(command, och)
+					}
+					close(och)
+					<-limit
+					return err
+				})
+			}
+		}
+
+		if err := eg.Wait(); err != nil {
+			log.Fatal(err)
+		}
+
+		o.Wait()
+		close(ch)
+	}()
+
+	for s := range ch {
+		fmt.Print(s)
+	}
+
+	return nil
+}
+
+func runCommand(command string, ch chan string) error {
+	var buf bytes.Buffer
+
+	fmt.Fprintf(&buf, "%s\n", command)
+
+	args := strings.Fields(command)
+	if runtime.GOOS == `windows` {
+		args[0] = convertToWindowsPath(args[0])
+	}
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdout = &buf
+	cmd.Stderr = &buf
+
+	err := cmd.Run()
+	if err != nil {
+		ch <- buf.String()
+		return err
+	}
+
+	ch <- buf.String()
+	return nil
+}
+
+func buildAndmd5sum(command string, ch chan string) error {
+	var buf bytes.Buffer
+
+	fmt.Fprintf(&buf, "%s\n", command)
+
+	args := strings.Fields(command)
+	envs := []string{}
+
+	ofs := 0
+	for _, f := range args {
+		spl := strings.SplitN(f, "=", 2)
+		if len(spl) == 2 {
+			envs = append(envs, f)
+			ofs++
+		} else {
+			break
+		}
+	}
+	args = args[ofs:]
+
+	if runtime.GOOS == `windows` {
+		args[0] = convertToWindowsPath(args[0])
+	}
+
+	output := ""
+	tmpOutput := ""
+	tmpdir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return err
+	}
+	for i := range args {
+		if args[i] == "-o" {
+			output = args[i+1]
+			ext := filepath.Ext(output)
+			tmpOutput = filepath.Join(tmpdir, fmt.Sprintf("%s.%s", filepath.Base(output), ext))
+			args[i+1] = tmpOutput
+		} else if strings.HasPrefix(args[i], "-o=") {
```

Comment:
Instead of doing command line parsing manually, I think you're much better off using [`flag.NewFlagSet`](https://pkg.go.dev/flag#NewFlagSet). This will not affect the command line flags parsed using `flag.Parse()`.

Reply:
I checked `flag.NewFlagSet()`.
If trying to parse with flagset, the `-size` and `-target` settings are required in addition to `-o`.
Is it possible to target only `-o`?

Reply:
Hmm. That's a fair point. Maybe the current approach is better.